no-downtime-upgrade/src/main/java/org/keycloak/playground/nodowntimeupgrade/naive_jackson/AbstractNaiveVersionedJacksonStorage.java

```diff
@@ -0,0 +1,135 @@
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+
+/**
+ * @param <M> Object model class
+ * @param <E> Entity class
+ * @author hmlnarik
+ */
+public abstract class AbstractNaiveVersionedJacksonStorage<M extends HasId<Integer>, E> extends AbstractMapStorage<M> {
+
+    private final static byte[] EMPTY = new byte[] {};
+
+    public static final ObjectMapper MAPPER = new ObjectMapper()
+      .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+      .configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false)
+
+      .setSerializationInclusion(JsonInclude.Include.NON_NULL)
+
+      .setVisibility(PropertyAccessor.ALL, Visibility.NONE)
+      .setVisibility(PropertyAccessor.FIELD, Visibility.ANY)
+    ;
+
+    private final int supportedVersion;
+
+    private final Function<M, E> modelToEntityFunc;
+
+    private final Class<E> entityClass;
+
+    private final Function<E, M> entityToModelFunc;
+
+    /**
+     *
+     * @param store Store where the objects are stored to.
+     * @param version Version of the object models this particular store is handling.
+     * @param modelToEntityFunc Function to create an entity from a model.
+     * @param entityClass Class representing the entity. Used for serialization from JSON to internal entity representation.
+     * @param entityToModelFunc Function to create a model from an entity that is guaranteed to be
+     *        at the model version compatible with {@code version}, migrated beforehand if necessary.
+     *        Entity version is compatible if it is either {@code supportedVersion} or {@code(#supportedVersion + 1)}.
+     */
+    public AbstractNaiveVersionedJacksonStorage(
+      ConcurrentMap<Integer, byte[]> store,
+      ModelVersion version,
+      Function<M, E> modelToEntityFunc,
+      Class<E> entityClass, Function<E, M> entityToModelFunc) {
+        super(store);
+        this.supportedVersion = version.getVersion();
+        this.modelToEntityFunc = modelToEntityFunc;
+        this.entityToModelFunc = entityToModelFunc;
+        this.entityClass = entityClass;
+    }
+
+    @Override
+    protected byte[] modelToByteArray(M object) throws Exception {
+        if (object == null) {
+            return EMPTY;
+        }
+        return MAPPER.writeValueAsBytes(modelToEntityFunc.apply(object));
+    }
+
+    @Override
+    protected M byteArrayToModel(byte[] bytes) throws Exception {
+        // validate input
+        if (bytes == null || bytes.length == 0) {
+            return null;
+        }
+
+        // validate logical entity version
+        ObjectNode tree = MAPPER.readValue(bytes, ObjectNode.class);
+        JsonNode ev = tree.get("entityVersion");
+        if (ev == null || ! ev.isInt()) {
+            return null;
+        }
+        int entityVersion = ev.asInt();
+
+        if (entityVersion < supportedVersion) {
+            tree = Migration.migrateTreeTo(entityVersion, supportedVersion, tree);
```

Comment:
Should we change also `tree.get("entityVersion")` to version it was migrated to so that we can later check for the current version of Json, for example, in `canReadVersion` function?

Reply:
`migrateTreeTo` guarantees that the tree is migrated to `supportedVersion`, so it is not necessary